### DISCUSSION TITLE
Process for large enhancements

### DISCUSSION THREAD

**Discussion by nrfox** [kiali/kiali#4208] [Governance]
Large enhancements or changes such as those proposed in the Roadmap Discussions will most likely require significant discussion, collaboration, and design work before the proposals are agreed upon and an implementation is accepted. Does Kiali currently have a process for proposing such changes similar to a [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md) (Kubernetes Enhancement Proposal) but perhaps less formal?

If not, I believe it would be good to formalize such a process and utilize it for driving some of the roadmap discussions forward. This could help move the discussions forward more quickly while also allowing for feedback along the way.

A few things that I think are worth including in such a process would be:

1. Clearly defined owner(s) of the enhancement. Their main role would be in shepherding the enhancement along but would not necessarily be responsible for implementation.
2. An open and thorough review process. Some background and context are given for the problem to go along with the proposed solution. Anyone can comment on or review the proposal before it is accepted.
3. A centralized place to store accepted proposals. Somewhere under the kiali org, perhaps the kiali/kiali repo.

Obviously not every enhancement or feature request should require this process but larger ones such as multicluster deployments could benefit.

**Comment by israel-hdez** [comment]
AFAIK, our current "process" is to propose a change in GitHub Discussions and iterate there. Before GHDiscussions, I once used a GDoc to make a proposal (in a similar way on how Istio makes proposals). See that doc here: https://docs.google.com/document/d/1Zc4QXEUm4wfa5u8NQv_8RdzQokwsfrOR7F8os_uGch4

TBH, I'm not sure if a formal process is going to help. The pain point is that is hard to engage people. See the "Roadmap" discussions. Last comment on that category was on July 12th (that was 11 days ago when I wrote this comment). We should find a way to engage people on new work before proposing any process.

In any case, I agree with all your 3 points. Currently, our centralized place for proposals is GitHub (either issues or discussions). But I think it's too limited.

**Reply by nrfox** [reply]
That google doc looks like a great proposal! My only qualm with gdocs in general is that I find them hard to search for and discover and some projects require you to sign in before viewing.

> TBH, I'm not sure if a formal process is going to help. The pain point is that is hard to engage people. See the "Roadmap" discussions. Last comment on that category was on July 12th (that was 11 days ago when I wrote this comment). We should find a way to engage people on new work before proposing any process.

That's a good point about engagement. Perhaps there would be more feedback if there was a more concrete proposal to go off of? GH discussions seem like a great way to start the conversation and gather general input. The next step would be something very similar to the gdoc you linked where the problem statement and proposed solutions are listed out in detail. This would give users something more concrete to review in order to see if the proposed solution met their requirements or even if the problem statement overlooks their requirements altogether. It's possible that even a formal process wouldn't increase engagement but it would at least provide a structure to move a proposal along without being stuck waiting for more feedback.

**Reply by israel-hdez** [reply]
> My only qualm with gdocs in general is that I find them hard to search for and discover and some projects require you to sign in before viewing.

I agree :-) 
I once saw a project using a GitHub repo for RFCs. That repo was just a bunch of markdown files with their proposals to implement (or already implemented) in the project. Then, people commented on PRs when adding RFCs or proposing new RFCs to replace older ones. That made reviewing a similar process to a code review.

Commenting just to add an option for storing proposals.

> GH discussions seem like a great way to start the conversation and gather general input. The next step would be something very similar to the gdoc you linked

I always like to see/give a proposal, so that there is something to start with. However, it also depends on the size of the "beast". I think a proposal should be implemented only by a few PRs (if not only one).

For example, the "Deployments model" discussion is too big to be covered in a single proposal. That one needs several ones.

> It's possible that even a formal process wouldn't increase engagement but it would at least provide a structure to move a proposal along without being stuck waiting for more feedback.

It depends... If the process includes a step that can block it, you still have that risk. For example, If the process includes a step that says that a proposal must be approved, it can get stuck.

**Reply by nrfox** [reply]
> I once saw a project using a GitHub repo for RFCs. That repo was just a bunch of markdown files with their proposals to implement (or already implemented) in the project. Then, people commented on PRs when adding RFCs or proposing new RFCs to replace older ones. That made reviewing a similar process to a code review.

That sounds great. I much prefer that format. I think that would make it easier to keep track of review comments and easier to find past and current proposals.

**Comment by abonas** [comment]
This is a very needed discussion. to the above points I'd add clearly stating the schedule/deadline for each discussion/decision to be "locked", to address the engagement/timing issues  (as stated above regarding when last comment was written)

**Reply by nrfox** [reply]
Agreed! Having a "window for comment" for discussions would be helpful in moving the proposal along while also providing an opportunity to give/receive feedback.

**Reply by israel-hdez** [reply]
May be we could use our already existent agile sprints?

I think a 3-week sprint is enough time for a discussion to evolve. Since Kiali is using GitHub issues and [GitHub Projects](https://github.com/orgs/kiali/projects), I think a discussion should be able to go from "Backlog" to "Done" in a sprint.

**Reply by jshaughn** [reply]
I agree that timelines help move discussions.  I think another thing is to limit the number of open discussions.   Having too may topics open at one time can be overwhelming, and people can shut down.  We may want to focus on one topic more intensely, including group calls, until it seems more well-defined and there is some agreement on how to move forward.

**Reply by nrfox** [reply]
> I agree that timelines help move discussions. I think another thing is to limit the number of open discussions. Having too may topics open at one time can be overwhelming, and people can shut down. We may want to focus on one topic more intensely, including group calls, until it seems more well-defined and there is some agreement on how to move forward.

Yes especially if there are timelines for how long a proposal is expected to be in a certain phase such as "in review" or "in design". At this point it would be unreasonable to expect 3-4 large proposals to be given a thorough review within a sprint. Perhaps a topic can stay "open for discussion" without a timeline associated with it but once an owner(s) has been assigned to that proposal, the expectation is that they will produce a design draft and others will review that draft within a reasonable time frame or else provide an update on progress. I like Edgar's suggestion of scoping to a sprint as that would provide a natural way to check in on the status of proposals during planning or to discuss them during the community meeting.

**Comment by lucasponce** [comment]
In general, we used the concept of "Epic" issue to define a large effort. An "Epic" could be splitted in multiple issues, and typically design discussions happen in the issue for visibility. Additional GDocs can be used if needed but the desire is to try to keep main flow in the issue to give better visibility (GDocs tend to populate of comments hard to follow, could be ok when participants are small).

Before to create an "Epic" we proposed to start with "discussions" as they have a more open format and they support better comments and visibility.

Note that a "discussion" can generate multiple "Epics" as well, if some area is clear and it has consensus is good to propose an "Epic" that can be taken into consideration for next iterations.

Also, note that the discussions we started are pretty big and open :) and they also define a long term strategy (which may involve new major versions of a project), so, with that in mind, creating a scheduling/timeline may not be appropiate, but yes, team leads/coordinators should work actively to get consensus and keep those discussions open and moving in the areas that are clear.

Another comment, when an issue is created, that doesn't mean that is added into a backlog, that typically happens after a grooming process.

**Reply by nrfox** [reply]
> In general, we used the concept of "Epic" issue to define a large effort. An "Epic" could be splitted in multiple issues, and typically design discussions happen in the issue for visibility. Additional GDocs can be used if needed but the desire is to try to keep main flow in the issue to give better visibility (GDocs tend to populate of comments hard to follow, could be ok when participants are small).
> 
> Before to create an "Epic" we proposed to start with "discussions" as they have a more open format and they support better comments and visibility.

I think that process can largely stay the same, I'm just advocating for more structure around the design discussion so that the process of going from github discussion --> design draft is more clear and it's more obvious which individuals are responsible for driving that train.

**Reply by lucasponce** [reply]
I think that the "design draft" shouldn't be a new forum.
I guess we are in the same page but perhaps it's just a matter of improve the details.

To me the natural flow should be:
- A discussion started, the discussion is open, it can touch an area which multiple enhancements.
- In the discussion, one goal should be to identify the areas that may have consensus, or the areas that may require more discussion.
- In the discussion, if needed, a design draft can be used, but only as an auxiliary asset (i.e. I need to bring some diagrams, or some flows, that in the discussion format are less handy), but participants should be responsible that the discussion "doesn't" move to another forum.
- When in the discussion some area is clear, it has consesus, and it may have all needed to work on it, a new issue can be created.
- "Epic" issues are used when the domain contains multiple tasks, and "Epic" may have one or several owners that are responsible to define the subtasks and help to coordinate.
- "Subtasks" of the "Epic" can be implemented worked by anyone, and the "Epic" owner may coordinate it according with the design.

I think this reflects most of the comments, if for the design, I'd say to use whatever public document that it may help, with the markdown I see the same limitations that happens in the issues; those seems good in big projects where there is a strong RFC process, but perhaps for a project like Kiali, it could be open (I'm thinking in diagrams, or pictures, as our UI nature is pretty high, so in that sense I prefer any document that supports comments), but linking the documents in the proper discussions / Epics and make sure that main discussions are kept in those.

Also, let's not forget that another possibility that we used in Kiali is the prototype, sometimes a pure "waterfall" model is not flexible enough, and when the domain is clear, there is an agreement to start a PoC in the way ot PR that can help as another way to test/validate the proposal.

A PoC may or may not be merged, that's not a bad thing (just to avoid the feeling if that happens), we have had several, and all PoC are good in that sense that they contribute to the design.

**Reply by nrfox** [reply]
It sounds like we might agree on the value of the process we just need to come to a consensus on the specific tooling.

> I think that the "design draft" shouldn't be a new forum.

My personal thoughts are that github discussions are a great way to have an informal discussion about a particular topic but it's harder to give specific feedback on a formal proposal using that tool. When reviewing a "design draft", you often want to give feedback on a specific aspect of it rather than the general idea. This is more like a code review where multiple people comment on specific sections and the authors respond to this feedback. Using github discussions, you would need to quote various aspects of the document in your comments and it'd be harder to track changes over time in terms of what feedback has been responded to by changes to the document itself by the author. Contrast this to a github PR where changes to the document itself are preserved and it's easy to comment on specific aspects of the design doc. Previous design documents for Kiali have used google docs which similarly allows for others to comment on specific sections easily but without the benefit of easily preserving changes to the document itself and being easily searchable and indexed within the kiali repo. With github PRs you can also easily vote or place a blocking hold on the design document itself.

> To me the natural flow should be:

    A discussion started, the discussion is open, it can touch an area which multiple enhancements.
    In the discussion, one goal should be to identify the areas that may have consensus, or the areas that may require more discussion.
    In the discussion, if needed, a design draft can be used, but only as an auxiliary asset (i.e. I need to bring some diagrams, or some flows, that in the discussion format are less handy), but participants should be responsible that the discussion "doesn't" move to another forum.
    When in the discussion some area is clear, it has consesus, and it may have all needed to work on it, a new issue can be created.
    "Epic" issues are used when the domain contains multiple tasks, and "Epic" may have one or several owners that are responsible to define the subtasks and help to coordinate.
    "Subtasks" of the "Epic" can be implemented worked by anyone, and the "Epic" owner may coordinate it according with the design.

Agreed on the flow I would just change the tooling or forum from all in discussions to moving the more formal discussions to github PRs, so something like:

- A discussion started, the discussion is open, it can touch an area which multiple enhancements.
- In the discussion, one goal should be to identify the areas that may have consensus, or the areas that may require more discussion.
- If an area requires more formal discussion ("this needs a KEP"), a design draft is created and submitted as a github PR.
- The PR serves as the place to give/receive feedback on that particular proposal until it is either accepted or rejected
- If accepted, new issues are created to implement it. This can be an epic or any number of issues required to implement.

The "design draft" is not a diagram with squares and lines but really just a way to formally summarize an idea and provide a forum for others to easily give feedback on specific aspects of it. That may or may not include diagrams.

I personally think it's worth moving the area of conversation around a topic from github discussions to a PR when there's a need for specific feedback because I think that's a better medium/forum for that kind of conversation. Whereas github discussions are great for more informal conversations around a topic. There may be a risk that topics become more fragmented as they are spread across both github discussions and PRs but it would also help establish what is actually being proposed.

All that being said, any KEP process is probably better than none and I'm more interested in establishing the process than in the specific tooling Kiali uses. The tooling can always be re-evaluated later if it's not working well.

**Reply by nrfox** [reply]
The recent change to the governance file is a great example of the process I'm proposing where informal discussion on a topic happens in github discussions [here](https://github.com/kiali/kiali/discussions/4199) and that transitions to a more formal proposal in the form of a github PR [here](https://github.com/kiali/kiali/pull/4311).

**Reply by lucasponce** [reply]
> a design draft is created and submitted as a github PR.

+1 on that, perhaps what we need is to have a repo or a folder to formalize new proposals/RFE that may require a specific design.

I guess to formalize and work in design, a simple .md is enough to iterate in a PR in the document, that may help to identify the next steps and put everyone in the same page.

**Reply by jshaughn** [reply]
+1, I think a /design dir, or something like that, of markdown docs that can be reviewed as a PR, sounds like a good idea.  I can never keep track of google docs, and this is better for exposing to community via github, like everything else.  And we can use Draft PRs while it is being prepared.

**Reply by nrfox** [reply]
I opened: https://github.com/kiali/kiali/pull/4581 as a template and to establish the process. Please take a look and add any comments/suggestions.
